### DISCUSSION TITLE
How to subclass and add functionality to `Bridgetown::Resource::Base`?

### DISCUSSION THREAD

**Discussion by joemasilotti** [bridgetownrb/bridgetown#676] [Q&A] [answered]
I'd like to add custom methods to my post resources. The [docs on resources](https://edge.bridgetownrb.com/docs/resources) explain that I can subclass `Bridgetown::Model::Base` and put a file in `plugins/` to get automatically picked up.

This doesn't seem to work. I'm assuming I'm missing something – can anyone help?

```ruby
# plugins/post.rb
class Post < Bridgetown::Resource::Base
  def foo
    "Hi!"
  end
end
```

```
# src/_pages/articles.erb
<% collections.posts.each do |post| %>
  <%= post.foo %>
<% end %>
```

```
[Bridgetown]   Conversion error: Bridgetown::Converters::ERBTemplates encountered an error while converting `_pages/articles.erb'
[Server] - Gracefully stopping, waiting for requests to finish
[Server] === puma shutdown: 2022-12-20 06:40:13 -0800 ===
[Server] - Goodbye!
[Bridgetown] Stopping auxiliary processes...
[Bridgetown]   Exception raised: NoMethodError
[Bridgetown] undefined method `foo' for #<Bridgetown::Resource::Base repo://posts.collection/_posts/2021-05-14-turbo-ios.md>
[Bridgetown]                  1: /Users/joemasilotti/workspace/projects/masilotti.com/bridgetown/src/_pages/articles.erb:17:in `block (2 levels) in __tilt_4300'
[Bridgetown]                  2: /Users/joemasilotti/workspace/projects/masilotti.com/bridgetown/src/_pages/articles.erb:16:in `each'
[Bridgetown]                  3: /Users/joemasilotti/workspace/projects/masilotti.com/bridgetown/src/_pages/articles.erb:16:in `block in __tilt_4300'
[Bridgetown]                  4: /Users/joemasilotti/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/bridgetown-core-1.2.0.beta4/lib/bridgetown-core/converters/erb_templates.rb:67:in `capture'
[Bridgetown]                  5: /Users/joemasilotti/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/view_component-2.74.1/lib/view_component/base.rb:300:in `content'
```

FWIW, the same thing occurs if I instead subclass `Bridgetown::Model::Base `.

**Comment by jaredcwhite** [comment, marked as answer]
So there's not really a concept of a Resource subclass. If you were using a Model subclass, you could get at that via `resource.model.some_method`. But there's a way to define resource extensions:

https://edge.bridgetownrb.com/docs/plugins/resource-extensions (1.2 docs)

That applies to any type of resource, so if you want different behavior depending on collection you'd handle that within the method.

Otherwise, at the model level, you could do:

```ruby
class Post < Bridgetown::Model::Base
  def foo
    "Hi!"
  end
end

# later:
post_resource.model.foo
```

**Reply by joemasilotti** [reply]
Oh, it's the model! Gotcha.

And the model doesn't have access to the resource anymore, right? So I can't access frontmatter/data on the resource in this subclass?

**Reply by jaredcwhite** [reply]
The model is the source of the raw front matter data, so it should all be there (aka `model.title`, `model.author`, etc.), other than a few computed values that occur when a resource is created based on the model.

**Reply by joemasilotti** [reply]
Hmm, maybe I'm missing something.

```
irb(main):011:0> site.collections.posts.resources.first.model.data
key `data' not found in attributes for new Post
nil
```

**Reply by joemasilotti** [reply]
Oh! If I access it via `data_attributes` it looks like it works. Is that intentional? Should there be an alias there or am I mixing two different concepts?

```
irb(main):001:0> site.collections.posts.resources.first.model.data_attributes.title
"Zero to App Store in 7 weeks"
```

**Reply by jaredcwhite** [reply]
The model _is_ the data essentially, so you can just use it like you use any ActiveModel. So `model.title`, not `model.data.title` or whatever

**Reply by joemasilotti** [reply]
Oh! Nice, that makes sense.

Does this data get all the usual `_defaults.yml` as well? I don't see those getting populated anywhere on the `Post` subclass.

**Reply by jaredcwhite** [reply]
Hmm… well there's no manipulation of data at the model layer, it's exactly how it was defined in the file system, or if it came from a CMS. So no defaults available yet. I think you could within your own model define "defaults" in a Ruby way by coding assignments inside of `initialize`

**Reply by joemasilotti** [reply]
Got it. Thanks Jared!

It sounds like I'm trying to abuse the model a bit too much. I think I'll stick with the global extension for now and throw a bunch of helpers in there.
